File: agenttel-python/src/agenttel/genai/openai.py
"""OpenAI SDK instrumentation wrapper."""

from __future__ import annotations

import functools
from typing import Any

from opentelemetry import context, trace

from agenttel.genai import attributes as genai_attr
from agenttel.genai.cost import ModelCostCalculator
from agenttel.genai.span_builder import GenAiSpanBuilder


def instrument_openai(client: Any = None) -> Any:
    """Instrument an OpenAI client for automatic tracing.

    Wraps chat.completions.create() (sync and async) to create
    gen_ai chat spans with full attribute instrumentation.

    Args:
        client: An OpenAI client instance. If None, patches the default client.

    Returns:
        The instrumented client.
    """
    if client is None:
        return client

    span_builder = GenAiSpanBuilder()
    _wrap_completions(client, span_builder)
    return client


def _wrap_completions(client: Any, span_builder: GenAiSpanBuilder) -> None:
    """Wrap the chat.completions.create method."""
    if not hasattr(client, "chat") or not hasattr(client.chat, "completions"):
        return

    original_create = client.chat.completions.create

    @functools.wraps(original_create)
    def traced_create(*args: Any, **kwargs: Any) -> Any:
        model = kwargs.get("model", "unknown")
        span = span_builder.start_chat_span(
            model=model,
            system="openai",
            temperature=kwargs.get("temperature"),
            max_tokens=kwargs.get("max_tokens"),
            top_p=kwargs.get("top_p"),
        )

        try:
            with trace.use_span(span, end_on_exit=False):
                response = original_create(*args, **kwargs)

                # Handle streaming
                if kwargs.get("stream"):
                    return _wrap_stream(response, span, span_builder, model)

                # Extract response data
                _finalize_span(span, span_builder, response, model)
                return response
        except Exception as exc:
            span.set_status(trace.Status(trace.StatusCode.ERROR, str(exc)))
            span.record_exception(exc)
            span.end()
            raise

    client.chat.completions.create = traced_create


def _finalize_span(
    span: trace.Span,
    span_builder: GenAiSpanBuilder,
    response: Any,
    model: str,
) -> None:
    """Extract response data and end span."""
    input_tokens = 0
    output_tokens = 0
    reasoning_tokens = 0
    cached_tokens = 0
    finish_reason = None
    response_id = None
    response_model = model

    if hasattr(response, "usage") and response.usage:
        input_tokens = getattr(response.usage, "prompt_tokens", 0) or 0
        output_tokens = getattr(response.usage, "completion_tokens", 0) or 0
        # Extended usage fields
        if hasattr(response.usage, "completion_tokens_details"):
            details = response.usage.completion_tokens_details
            if details:
                reasoning_tokens = getattr(details, "reasoning_tokens", 0) or 0
        if hasattr(response.usage, "prompt_tokens_details"):
            details = response.usage.prompt_tokens_details
            if details:
                cached_tokens = getattr(details, "cached_tokens", 0) or 0

    if hasattr(response, "id"):
        response_id = response.id
    if hasattr(response, "model"):
        response_model = response.model
    if hasattr(response, "choices") and response.choices:
        finish_reason = getattr(response.choices[0], "finish_reason", None)

    span_builder.end_span_with_response(
        span=span,
        model=response_model,
        input_tokens=input_tokens,
        output_tokens=output_tokens,
        reasoning_tokens=reasoning_tokens,
        cached_tokens=cached_tokens,
        finish_reason=finish_reason,
        response_id=response_id,
    )


def _wrap_stream(
    stream: Any,
    span: trace.Span,
    span_builder: GenAiSpanBuilder,
    model: str,
) -> Any:
    """Wrap a streaming response to capture final usage."""
    input_tokens = 0
    output_tokens = 0
    finish_reason = None
    response_model = model

    def _gen():
        nonlocal input_tokens, output_tokens, finish_reason, response_model
        try:
            for chunk in stream:
                if hasattr(chunk, "model") and chunk.model:
                    response_model = chunk.model
                if hasattr(chunk, "usage") and chunk.usage:
                    input_tokens = getattr(chunk.usage, "prompt_tokens", 0) or 0
                    output_tokens = getattr(chunk.usage, "completion_tokens", 0) or 0
                if hasattr(chunk, "choices") and chunk.choices:
                    fr = getattr(chunk.choices[0], "finish_reason", None)
                    if fr:
                        finish_reason = fr
                yield chunk
        finally:
            span_builder.end_span_with_response(
                span=span,
                model=response_model,
                input_tokens=input_tokens,
                output_tokens=output_tokens,
                finish_reason=finish_reason,
            )

    return _gen()

File: agenttel-node/src/genai/openai.ts
/**
 * OpenAI SDK instrumentation wrapper.
 *
 * Monkey-patches `client.chat.completions.create()` to automatically create
 * OpenTelemetry spans with GenAI semantic attributes for every chat completion.
 *
 * Usage:
 *   import OpenAI from 'openai';
 *   import { instrumentOpenAI } from '@agenttel/node';
 *
 *   const client = instrumentOpenAI(new OpenAI());
 *   const response = await client.chat.completions.create({ model: 'gpt-4o', ... });
 */

import { type Span } from '@opentelemetry/api';
import { GenAiSpanBuilder, type ChatResult } from './span-builder.js';

const SYSTEM = 'openai';

/**
 * Instrument an OpenAI client for automatic tracing.
 *
 * Wraps `chat.completions.create()` to create gen_ai chat spans with
 * full attribute instrumentation including token usage, cost, and streaming.
 *
 * @param client - An OpenAI client instance.
 * @returns The instrumented client (same reference, mutated in place).
 */
export function instrumentOpenAI<T extends Record<string, any>>(client: T): T {
  if (!client) return client;

  const chat = client.chat;
  if (!chat?.completions?.create) return client;

  const spanBuilder = new GenAiSpanBuilder();
  const originalCreate = chat.completions.create.bind(chat.completions);

  chat.completions.create = async function tracedCreate(...args: any[]) {
    const params = args[0] ?? {};
    const model: string = params.model ?? 'unknown';

    const span = spanBuilder.startChatSpan(model, SYSTEM);

    // Set optional request parameters.
    GenAiSpanBuilder.setRequestParams(
      span,
      params.temperature ?? undefined,
      params.max_tokens ?? params.max_completion_tokens ?? undefined,
      params.top_p ?? undefined,
    );

    try {
      const response = await originalCreate(...args);

      // Handle streaming responses.
      if (params.stream) {
        return wrapStream(response, span, model);
      }

      // Non-streaming: extract response data and finalize span.
      finalizeSpan(span, response, model);
      return response;
    } catch (err) {
      GenAiSpanBuilder.endSpanError(span, err instanceof Error ? err : new Error(String(err)));
      throw err;
    }
  };

  return client;
}

/**
 * Extract response data and end the span for a non-streaming response.
 */
function finalizeSpan(span: Span, response: any, requestModel: string): void {
  let inputTokens = 0;
  let outputTokens = 0;
  let reasoningTokens = 0;
  let cachedTokens = 0;
  let finishReason: string | undefined;
  let responseId: string | undefined;
  let responseModel = requestModel;

  if (response.usage) {
    inputTokens = response.usage.prompt_tokens ?? 0;
    outputTokens = response.usage.completion_tokens ?? 0;

    // Extended usage fields.
    if (response.usage.completion_tokens_details) {
      reasoningTokens = response.usage.completion_tokens_details.reasoning_tokens ?? 0;
    }
    if (response.usage.prompt_tokens_details) {
      cachedTokens = response.usage.prompt_tokens_details.cached_tokens ?? 0;
    }
  }

  if (response.id) responseId = response.id;
  if (response.model) responseModel = response.model;
  if (response.choices?.length > 0) {
    finishReason = response.choices[0].finish_reason ?? undefined;
  }

  // Set extended token details.
  if (reasoningTokens > 0 || cachedTokens > 0) {
    GenAiSpanBuilder.setTokenDetails(span, {
      reasoningTokens: reasoningTokens || undefined,
      cachedReadTokens: cachedTokens || undefined,
    });
  }

  const result: ChatResult = {
    responseModel,
    responseId: responseId ?? '',
    inputTokens,
    outputTokens,
    finishReasons: finishReason ? [finishReason] : undefined,
  };

  GenAiSpanBuilder.endSpanSuccess(span, result);
}

/**
 * Wrap a streaming response (async iterable) to accumulate token usage
 * across chunks and end the span when the stream completes.
 */
function wrapStream(stream: any, span: Span, requestModel: string): AsyncIterable<any> {
  let inputTokens = 0;
  let outputTokens = 0;
  let finishReason: string | undefined;
  let responseModel = requestModel;

  const originalIterator = stream[Symbol.asyncIterator]
    ? stream[Symbol.asyncIterator].bind(stream)
    : null;

  if (!originalIterator) {
    // Not an async iterable, end the span and return as-is.
    GenAiSpanBuilder.endSpanSuccess(span, {
      responseModel,
      responseId: '',
      inputTokens: 0,
      outputTokens: 0,
    });
    return stream;
  }

  const wrappedIterable: AsyncIterable<any> = {
    [Symbol.asyncIterator]() {
      const iterator = originalIterator();
      return {
        async next() {
          try {
            const result = await iterator.next();
            if (result.done) {
              // Stream complete: finalize span with accumulated data.
              GenAiSpanBuilder.endSpanSuccess(span, {
                responseModel,
                responseId: '',
                inputTokens,
                outputTokens,
                finishReasons: finishReason ? [finishReason] : undefined,
              });
              return result;
            }

            const chunk = result.value;

            // Accumulate data from stream chunks.
            if (chunk.model) responseModel = chunk.model;
            if (chunk.usage) {
              inputTokens = chunk.usage.prompt_tokens ?? inputTokens;
              outputTokens = chunk.usage.completion_tokens ?? outputTokens;
            }
            if (chunk.choices?.length > 0) {
              const fr = chunk.choices[0].finish_reason;
              if (fr) finishReason = fr;
            }

            return result;
          } catch (err) {
            GenAiSpanBuilder.endSpanError(
              span,
              err instanceof Error ? err : new Error(String(err)),
            );
            throw err;
          }
        },
      };
    },
  };

  // Preserve any non-iterator properties from the original stream (e.g., controller, response).
  return Object.assign(wrappedIterable, stream);
}
